Abstract the core module (Closes #97)

szareiangm opened the pull request:
This pull request is focused on having anything related to Elasticsearch layer in `elasticsearch` module.  It is mostly moving things around to re-organize, i.e.:  Like renaming `ElasticesarchEmitter` to `Emitter` or `KinesisElasticsearchPipeline` to `KinesisPipeline`.

Two side-notes: 
- There is a `partition` notion in JsonRecord that will be used in the next pull request that I will make, but I kept it here just as a placeholder. I just `index` name as it value since all of the data goes to one index now.
- Also, `document_type` in Elastic search is getting deprecated, but I kept it in `elasticsearch` module.

CC: @Nafid

snowplowcla commented:
@szareiangm has signed the Software Grant and Corporate Contributor License Agreement

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,190 @@
+  @throws[IOException]
+  def emit(buffer: UnmodifiableBuffer[EmitterJsonInput]): List[EmitterJsonInput] = {
+    attemptEmit(buffer.getRecords)
+  }
```
nit: we can remove the brackets

szareiangm replied:
addressed

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,190 @@
+        case _ => emit(validRecords)
+      }
+
+      (invalidRecords ++ rejects).asJava
```
why convert to java's list?

szareiangm replied:
`emit` needs Java list

szareiangm replied:
`emit` needs Java type for its return value and this class mainly implements KCL emitter for Java.

BenFradet replied:
ah, can you add `override` to the methods which are from `IEmitter` that will make things clearer

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,190 @@
+import scala.collection.JavaConverters._
+
+
+import sinks.ISink
```
imports need to be sorted, you can look at the existing code for guidelines

szareiangm replied:
addressed

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,190 @@
+    * @param records List of records to send to Elasticsearch
+    * @return List of inputs which Elasticsearch rejected
+    */
+  def emit(records: SList[EmitterJsonInput]): SMBuffer[EmitterJsonInput] = {
```
we shouldn't expose mutable data structures

szareiangm replied:
`private def emit` solves it. right?

szareiangm replied:
solved. This method is not exposed. Instead, emit is exposed and that one is good: `emit(buffer: UnmodifiableBuffer[EmitterJsonInput]): List[EmitterJsonInput]`

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala:
```diff
@@ -0,0 +1,116 @@
+
+    streamLoaderConf
+  }
+  val config:StreamLoaderConfig = parseConfig().get
```
please don't use `get` on options

szareiangm replied:
But it used to be the same as I did but in the child class: https://github.com/snowplow/snowplow-elasticsearch-loader/blob/7861698a6da7fb309b0b7f4c7b51266d1b93a1c3/elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/ElasticsearchHTTPSinkApp.scala#L28

I just moved it to have the variable in the trait. What is your suggestion?

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala:
```diff
@@ -0,0 +1,116 @@
+
+    val goodSink = config.sink.good match {
+      case "stdout" => Some(new StdouterrSink)
+      case "elasticsearch" => None
```
?

szareiangm replied:
it used to be like this from before:  https://github.com/snowplow/snowplow-elasticsearch-loader/blob/7861698a6da7fb309b0b7f4c7b51266d1b93a1c3/core/src/main/scala/com.snowplowanalytics.stream/loader/ElasticsearchSinkApp.scala#L150

Because the default sink is `elasticsearch` and goodSink is the extra addition.

szareiangm commented:
@BenFradet I think I addressed all of your actionable comments. Please go for the next round of review if needed.

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,181 @@
+import clients.BulkSender
+
+/**
+  * ElasticsearchEmitter class for any sort of BulkSender Extension
```
wrong doc

szareiangm commented:
@BenFradet I have no more comments to be addressed. Any chance for this PR to be merged?

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,179 @@
+    * @param records List of records to send
+    * @return List of inputs which the sink rejected
+    */
+  def emit(records: SList[EmitterJsonInput]): SMBuffer[EmitterJsonInput] = {
```
can we not have a buffer here?

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,179 @@
+    */
+   def fail(records: List[EmitterJsonInput]): Unit = {
+    records.asScala.foreach {
+      _ match {
```
no need for that

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala:
```diff
@@ -0,0 +1,117 @@
+  private def parseConfig(): Option[StreamLoaderConfig] = {
+    val projectName = "snowplow-stream-loader"
+    case class FileConfig(config: File = new File("."))
+    val parser:OptionParser[FileConfig] = new scopt.OptionParser[FileConfig](projectName) {
```
missing space

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala:
```diff
@@ -0,0 +1,117 @@
+/**
+ * Main entry point for the Elasticsearch sink
+ */
+trait StreamLoaderApp extends App {
```
could we move away from using `App`?

szareiangm replied:
This is from the current code. If I have done it in this PR, you would have added to it the too much work list :stuck_out_tongue_winking_eye: Also, as you know `App` has internal mechanisms that removing it, might not look straight-forward and with small change.

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala:
```diff
@@ -0,0 +1,117 @@
+    streamLoaderConf
+  }
+
+  val config:StreamLoaderConfig = parseConfig().get
```
please don't use `get`

szareiangm replied:
The same used to be in the code. I just moved it. 
https://github.com/snowplow/snowplow-elasticsearch-loader/pull/100/files#diff-20d604dd7232c1b254b18a7fc7440f04L28

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/model.scala:
```diff
@@ -18,9 +18,7 @@
 package com.snowplowanalytics.stream.loader
-
 import java.text.SimpleDateFormat
-
```
can you leave those lines

BenFradet commented on elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/ElasticsearchStreamLoaderApp.scala:
```diff
@@ -0,0 +1,125 @@
+        config.elasticsearch.get.client.password,
+        config.elasticsearch.get.client.maxTimeout,
+        config.elasticsearch.get.client.maxRetries,
+        config.elasticsearch.get.cluster.documentType,
```
`get` shouldn't be needed, you case should be `case Some(esConfig)` => esConfig.client.endpoint`

BenFradet commented on elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala:
```diff
@@ -0,0 +1,166 @@
+    restClientBuilder.setDefaultHeaders(headers)
+  }
+
+  private val client: HttpClient = HttpClient.fromRestClient(restClientBuilder.build())
```
why was the client instantiation changed?

szareiangm replied:
a mistake! reverting this part.

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/StreamLoaderApp.scala:
```diff
@@ -0,0 +1,117 @@
+
+  val goodSink = config.sink.good match {
+    case "stdout" => Some(new StdouterrSink)
+    case "elasticsearch" => None
```
This can't be part of the core since it's what you're trying to abstract over.

You can look at [stream-enrich](https://github.com/snowplow/snowplow/tree/master/3-enrich/stream-enrich) to know how we've abstracted over the streaming platform.

szareiangm replied:
I will move it `es` module since core doesn't need for any reason yet. Having it like `stream-enrich` is a long shot for this ticket.

BenFradet commented:
>I wish we would have a styling config file for this project.

Yup, that's a very good point, we've been slowly integrating scalafmt in other projects, I created #101.

>Transformers needed to be abstracted because they generate ElasticsearchObject. It could have been done in the es module

Agreed but they didn't need to be moved in this commit? If you look at the diff for this PR, almost half is related to transformers and that should just be a few lines. Do you see what I mean?

szareiangm commented:
> Agreed but they didn't need to be moved in this commit? If you look at the diff for this PR, almost half is related to transformers and that should just be a few lines. Do you see what I mean?

You are right. I will more cautious about amount of changes in each PR for next time. I should have broken down this issue, but now is a bit late. 😞  I addressed your comments except two (`good sink` change and eliminating `App`), that I wanted to defer for next pull requests since they might involve too much work for this PR.

BenFradet commented on core/src/test/scala/com.snowplowanalytics.stream.loader/PlainJsonTransformerSpec.scala:
```diff
@@ -12,23 +12,35 @@
+      val input = """{"key1":"value1","key2":"value2","key3":"value3"}"""
+
+      val result = new PlainJsonTransformer("snowplow")
+        .fromClass(input -> JsonRecord(parse(input).asInstanceOf[JObject], "snowplow").success)
```
why is the `asInstanceOf` needed?

szareiangm replied:
I found an example of converting JValue to JObject like this,  although I knew `asInstanceOf` is not your favorite. Let me check alternatives.

BenFradet commented on elasticsearch/src/test/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSenderSpec.scala:
```diff
@@ -10,54 +10,64 @@
-      sender.sendToElasticsearch(data) must_== List.empty
+
+      val validInput: EmitterJsonInput = "good" -> JsonRecord(
+        parse("""{"s":"json"}""").asInstanceOf[JObject],
```
same

szareiangm replied:
removed

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/JsonRecord.scala:
```diff
@@ -18,10 +18,12 @@
+ * @param partition partition to send the data to, currently not used.
  */
-case class JsonRecord(json: String, id: Option[String])
+case class JsonRecord(json: JObject, partition: String)
```
can we move this out of this PR, this doesn't seem related to abstracting the core module

szareiangm replied:
removed it to have it with partition pr

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala:
```diff
@@ -41,68 +38,53 @@ import com.snowplowanalytics.snowplow.analytics.scalasdk.json.EventTransformer._
+    extends ITransformer[ValidatedJsonRecord, EmitterJsonInput]
     with StdinTransformer {
 
+  val log = LoggerFactory.getLogger(getClass)
```
is this used?

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/JsonTransformer.scala:
```diff
@@ -41,68 +38,53 @@ import com.snowplowanalytics.snowplow.analytics.scalasdk.json.EventTransformer._
  */
-class SnowplowElasticsearchTransformer(documentIndex: String, documentType: String)
-    extends ITransformer[ValidatedRecord, EmitterInput]
+class JsonTransformer(documentIndex: String)
```
I think the rename is confusing wrt the `PlainJsonTransformer`, maybe `EnrichedEventJsonTransformer` wdyt?

szareiangm replied:
but all of them are `EnrichedEventJsonTransformer`

szareiangm replied:
renamed, although I am not fan of long names. you are the boss

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala:
```diff
@@ -35,11 +39,10 @@ import Scalaz._
- * @param the elasticsearch index type
+ * @param documentIndex elasticsearch index name
  */
-class BadEventTransformer(documentIndex: String, documentType: String)
```
ideally, the `documentType` removal should have been in its own commit

szareiangm replied:
reverted.

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/BadEventTransformer.scala:
```diff
@@ -59,14 +64,15 @@ class BadEventTransformer(documentIndex: String, documentType: String)
-  override def fromClass(record: ValidatedRecord): EmitterInput =
-    (record._1, record._2.map(j => new ElasticsearchObject(documentIndex, documentType, j.json)))
+  override def fromClass(record: ValidatedJsonRecord): EmitterJsonInput =
+    (record._1, record._2.map(j => j))
```
why is the map needed?

szareiangm replied:
simplified.

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/model.scala:
```diff
@@ -74,10 +74,11 @@ package model {
     username: Option[String],
     password: Option[String],
     maxTimeout: Long,
+    maxRetries: Int,
```
this should have also been in its own commit

szareiangm replied:
removed

BenFradet commented on core/src/test/scala/com.snowplowanalytics.stream.loader/BadEventTransformerSpec.scala:
```diff
@@ -28,13 +35,11 @@ class BadEventTransformerSpec extends Specification {
-      elasticsearchObject.getType must_== "bad"
-      elasticsearchObject.getSource must_== input
+      val result = new BadEventTransformer("snowplow")
+        .fromClass(input -> JsonRecord(parse(input).asInstanceOf[JObject], null).success)
```
same with the `asInstanceOf` please don't use them

szareiangm replied:
removed.

BenFradet commented on core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/PlainJsonTransformer.scala:
```diff
@@ -75,8 +73,11 @@ class PlainJsonTransformer(documentIndex: String, documentType: String)
-      case None         => "Json parsing error".failureNel
+      case Some(jvalue) =>
+        JsonRecord(
+          jvalue.asInstanceOf[JObject] ~ ("id" -> UUID.randomUUID().toString),
```
asInstanceOf

szareiangm replied:
Json4s is a nice library but its inheritance kills! changed.

BenFradet commented on elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/clients/ElasticsearchBulkSender.scala:
```diff
@@ -80,23 +80,28 @@ class ElasticsearchSenderHTTP(
   override def close(): Unit = ()
 
-  override def sendToElasticsearch(records: List[EmitterInput]): List[EmitterInput] = {
+  def send(records: List[EmitterJsonInput]): List[EmitterJsonInput] = {
```
can you add an override keyword

szareiangm replied:
done

szareiangm commented:
Addressed the comments. Another round of review if I may ask please? @BenFradet

BenFradet commented:
sorry I couldn't give another review sooner, I'll have a look this afternoon :+1:

szareiangm commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,175 @@
+   * @param records List of records to send
+   * @return List of inputs which the sink rejected
+   */
+  def emit(records: SList[EmitterJsonInput]): SList[EmitterJsonInput] = {
```
done

szareiangm replied:
`NsqSourceExecutor` uses it

szareiangm replied:
no. `NsqSourceExecutor` uses this one

szareiangm commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,175 @@
+   * @param recordLimit emitter record limit
+   * @returns a list of buffers
+   */
+  def splitBuffer(
```
done

szareiangm commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,175 @@
+  /**
+   * Closes the Sink client when the KinesisConnectorRecordProcessor is shut down
+   */
+  def shutdown(): Unit = bulkSender.close
```
done

szareiangm commented on core/src/main/scala/com.snowplowanalytics.stream/loader/Emitter.scala:
```diff
@@ -0,0 +1,175 @@
+   *
+   * @param records List of failed records
+   */
+  def fail(records: List[EmitterJsonInput]): Unit = {
```
override added.

szareiangm commented on core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/EnrichedEventJsonTransformer.scala:
```diff
@@ -41,68 +37,52 @@ import com.snowplowanalytics.snowplow.analytics.scalasdk.json.EventTransformer._
-        case None     => new ElasticsearchObject(documentIndex, documentType, r.json)
-    }))
+  override def fromClass(record: ValidatedJsonRecord): EmitterJsonInput =
+    record.map(_.map(r => r))
```
replaced with `record`

szareiangm commented on core/src/main/scala/com.snowplowanalytics.stream/loader/transformers/PlainJsonTransformer.scala:
```diff
@@ -62,10 +62,8 @@ class PlainJsonTransformer(documentIndex: String, documentType: String)
-    record.map(_.map(r => new ElasticsearchObject(documentIndex, documentType, uuid, r.json)))
-  }
+  override def fromClass(record: ValidatedJsonRecord): EmitterJsonInput =
+    record.map(_.map(r => r))
```
replaced with `record`

szareiangm commented on core/src/test/scala/com.snowplowanalytics.stream.loader/EmitterSpec.scala:
```diff
@@ -40,46 +41,51 @@ import org.specs2.mutable.Specification
+        def send(records: List[EmitterJsonInput]): List[EmitterJsonInput] = List.empty
+        def close(): Unit                                                 = ()
+        def logHealth(): Unit                                             = ()
+        val tracker                                                       = None
```
added `override` to all

szareiangm commented on elasticsearch/src/main/scala/com/snowplowanalytics/stream/loader/ElasticsearchStreamLoaderApp.scala:
```diff
@@ -0,0 +1,145 @@
+          case Bad =>
+            new BadEventTransformer(
+              config.elasticsearch.get.cluster.index,
+              config.elasticsearch.get.cluster.documentType)
```
I have moved it to `match`

szareiangm commented:
addressed the comments @BenFradet . please check

szareiangm commented:
Yay! Send me a Snowplow T-Shirt. I should wear it in the office 🙌 😋

alexanderdean commented:
> Send me a Snowplow T-Shirt.

I've created an email thread with Nicki on our side to get that arranged!
